### ISSUE TITLE
Provide a way to simplify writing mixed with

### ISSUE THREAD

**Issue by K-Lovelace** [neo4j-contrib/cypher-dsl#190]
Hello,

We have been struggling a bit with readability lately.

We have a lot of with written like this:

`
            query.with(u.getRequiredSymbolicName(), uId.getRequiredSymbolicName(), r.getRequiredSymbolicName(),
                  app.getRequiredSymbolicName(), caseExpression().when(ll.isNull())
                      .then(l.getRequiredSymbolicName())
                      .elseDefault(ll.getRequiredSymbolicName())
                      .as(l.getRequiredSymbolicName()))
`

It's very hard to read because of all the getRequiredSymbolicName. We would like a way to write: 

`
            query.with(u, uId, r, app, caseExpression().when(ll.isNull())
                      .then(l)
                      .elseDefault(ll)
                      .as(l))
`

Which would be much easier to read. (u, uld, r, app are Nodes, which do not mix well with AliasedExpressions).

Maybe there already is a way ? 

Thanks for the big 2021.2.0 update, it's getting nicer and nicer to use!

**Comment by michael-simons** [comment]
Yeah, I do hear you loud and clear, I am not happy with that either. I am gonna put up my thinking hat for a bit.

**Comment by michael-simons** [comment]
Hi again, can you please checkout the main branch and built it to see whether it suites your needs now? 

You'll need JDK 11 (`./mvnw clean verify -pl \!org.neo4j:neo4j-cypher-dsl-native-tests`).

You can also get a snapshot here: https://jitpack.io/#neo4j-contrib/cypher-dsl/main-SNAPSHOT 
However, if you using it with Spring Data Neo4j, you would need to exclude the Cypher-DSL from the original coordinates.

Also, an explicit thank you for your kind words, much appreciated!

**Comment by K-Lovelace** [comment]
Hi, i haven't tested yet but I have a colleague that will definetely check it out today. 

But reading your commit I think it's exactly what we've been looking for! It will simplify soooo much code. 

I wonder if you could add it also for other methods, like `as` or `caseExpression.then()` (I think there are some other use cases but I can't think of them as I'm not coding with cypherdsl right now). 

There also is something that's been in a corner of my head for some time: To describe our model and avoid having string literals everywhere, we use enums like this:

```
enum UserProperty {
     firstname,
     lastname
}
```

and then use them like this:

```
user.property(UserProperty.firstname.name())
```

which works perfectly but the added `name()` call really make the code header to read. Is there an intended way of doing that ? Or is it possible to have something like an interface to implement to be allowed to pass enum constants directly to methods awaiting strings as arguments ?

**Comment by michael-simons** [comment]
Yeah there are indeed a couple of other places, I was thinking more about `return`. That will be even harder to do in a way that is API compatible. With the change above I already will break some scenarios in which now an explicit cast is required.

You can indeed have something like a static model, please have a look here:
https://github.com/neo4j-contrib/cypher-dsl/tree/main/neo4j-cypher-dsl-examples/neo4j-cypher-dsl-examples-core/src/test/java/org/neo4j/cypherdsl/examples/model

http://neo4j-contrib.github.io/cypher-dsl/current/#static-meta-model

While we did provide examples for SDN6, it's basically agnostic, as the example code shows.

If you are willing to try this out, I would be super happy to assist.

**Comment by K-Lovelace** [comment]
Wow, totally missed this feature. So how would it work if, for brevity, we would like to have a supermodel having several very common properties, like name or uuid, to distribute with every other models ? Simple inheritance would work ? Same question for common relations between Models ?

**Comment by michael-simons** [comment]
Let me see if I can bring up an example.

**Comment by K-Lovelace** [comment]
That would be nice, it would be the founding stone for our Spring-based model description so we are really into this feature. I'll make a few tries myself waiting for your input.

**Comment by michael-simons** [comment]
Have a look: https://github.com/neo4j-contrib/cypher-dsl/commit/8b6f030a230284975f9033059921ccf51de40ede 

That's how I probably model it.
Note that I added a parameter to the relationship constructor for adding additional types.

If you try out the code generator and ran into issue, please report them! We would love to move the annotation processor forward.

**Comment by K-Lovelace** [comment]
Very nice examples, exactly what I needed.

Well, I did try it, and will not be using it for now, here's why: 
There's no support for inheritance (and we have lots of them, not only for common properties like uuid or name, but mostly for specific nodes)
I could'nt make it work properly, every time i added it and built my maven project with it, it would work, and then on the next build it would crash. I would then have to remove the plugin, build again, delete target, add the plugin, then rebuild... I guess something was wrong with the way I was doing it so i decided to code myself for now and to check it out again later. 

Another question, how would you use a relation (:Object)-[:IN]->(anything) ? I can write the class just fine but then you cannot have a static ready-to-use implementation can you ? if I have a room = anyNode("room") and object = Object.OBJECT.named("object"), I should do match(new In(object, room)) ? Or is there a way to write this more "cypherly" ?

**Comment by K-Lovelace** [comment]
Ooooh I think I found something funny:

If I use `Division.DIVISION.named("d")`, the signature of the method does not return `Division` but `AbstractNodeDefinition<Division>` and if Division has a property defined, (e.g. not in AbstractNodeDefition), I cannot use it. 

For now I use a workaround by using the named(SymbolicName) method but how would you get around that ?

**Comment by K-Lovelace** [comment]
Also, why is it that when I specify no name, a random name is assigned ? I don't what those kind of names in my generated cyphers, and right now I'm writing node(Division.$TYPE) to avoid that and it's less readable in comparison to the awesome Division.DIVISION that I would like to use.

**Comment by K-Lovelace** [comment]
Any news on my last question ? I think I found some ways over the others but this last one (automatically generated ones) actually gave us a hard time, where a node generated using nodeA.relationshipTo(node("Label")).asCondition() was getting an automatically generated name and therefore making the request fail. Is there a way to disable that ? I understand the use cases but we prefer giving explicit names when and if we need them.

**Comment by michael-simons** [comment]
> I could'nt make it work properly, every time i added it and built my maven project with it, it would work, and then on the next build it would crash. 

if you have a reproducer for a crash that would be awesome

> If I use Division.DIVISION.named("d"), the signature of the method does not return Division but AbstractNodeDefinition<Division> and if Division has a property defined, (e.g. not in AbstractNodeDefition), I cannot use it

Excellent catch. There is no other way right now than your workaround. 

> Also, why is it that when I specify no name, a random name is assigned ?

You can always name the static model. We need the names to mach things. See the related discussion here: https://github.com/neo4j-contrib/cypher-dsl/issues/70

**Comment by michael-simons** [comment]
Btw you should be able to cast things into the desired type: `Division d = (Division) Division.DIVISION.named("x");`.

**Comment by michael-simons** [comment]
I improved the example. Generics, always a pleasure and I still manage to mess them up after all these years:

https://github.com/neo4j-contrib/cypher-dsl/commit/2d2b8132730e9d1048a016a3f3f0c49b6d8e831c 

Now the methods (not only `named`) give you the expected type.

**Comment by nazeeh-alsaifi** [comment]
Thank you @michael-simons for the example.

> Have a look: [8b6f030](https://github.com/neo4j/cypher-dsl/commit/8b6f030a230284975f9033059921ccf51de40ede)
> 
> That's how I probably model it. Note that I added a parameter to the relationship constructor for adding additional types.
> 
> If you try out the code generator and ran into issue, please report them! We would love to move the annotation processor forward.

in my case I am using SDN 6+ annotation processor to create the metamodels. is there any way to take parent class properties into consideration in this case?
